drivers/overlay/overlay.go

```diff
@@ -2020,11 +2021,27 @@ func (d *Driver) getWhiteoutFormat() archive.WhiteoutFormat {
 }
 
 type overlayFileGetter struct {
-	diffDirs []string
+	diffDirs        []string
+	composefsMounts map[string]*os.File // map from diff dir to the directory with the composefs blob mounted
 }
 
 func (g *overlayFileGetter) Get(path string) (io.ReadCloser, error) {
+	buf := make([]byte, unix.PathMax)
 	for _, d := range g.diffDirs {
+		if f, found := g.composefsMounts[d]; found {
+			// there is no *at equivalent for getxattr, but it can be emulated by opening the file under /proc/self/fd/$FD/$PATH
+			len, err := unix.Getxattr(fmt.Sprintf("/proc/self/fd/%d/%s", int(f.Fd()), path), "trusted.overlay.redirect", buf)
```

Comment:
This `Getxattr` probably wants at some point to be factored out to a helper function, we have a *lot* of copies of the `/proc/self/fd/` trick in this codebase, and there is actually finally some movement on a proper syscall for `lgetxattrat` that we could change in one place instead of many.

---

On a different topic it feels a bit like we're poking into composefs internals here?  One thing I've thought about in the past that would be nice perhaps is something like an `ioctl()` usable on an overlayfs that would reopen the file descriptor with the real lower file.  Specifically, what I hit is I wanted to be able to reflink across an overlay view and a target in the real underlying filesystem.  That's a case that it seems to me we could make safe and viable, and in that model the overlayfs kernel implementation would handle "chasing" this `trusted.overlay.redirect`.